File: web/src/lib/account/tours.ts
/**
 * The account's tour progress, over the API.
 *
 * Used from the SERVER only — Astro frontmatter and the endpoint under
 * pages/api. The bearer token in this product never reaches the browser:
 * every editorial screen builds its client in frontmatter
 * (`createEditorialApi(API_BASE_URL, session.token)`) and renders the
 * result, and a guided tour is not a reason to start handing tokens to
 * page scripts. So the browser reads its starting cursor from an attribute
 * the server rendered, and writes through a same-origin endpoint that
 * holds the session on this side of the wire.
 *
 * Same seam as the other clients: `baseUrl` empty means no API is
 * configured, and every call degrades to "no progress recorded" rather
 * than throwing. A tour is the least important thing on any of these
 * screens, and it must never be the reason one fails to render.
 */

/** A tour id to its cursor: a step index as text, or 'done'. */
export type TourProgress = Readonly<Record<string, string>>;

/** Nothing recorded — the answer whenever the API cannot be asked. */
export const NO_PROGRESS: TourProgress = Object.freeze({});

/** What the API accepts as a cursor; mirrors the Go handler's pattern. */
const CURSOR = /^(done|[0-9]{1,4})$/;

/** What the API accepts as a tour id; mirrors the Go handler's pattern. */
const TOUR_ID = /^[a-z][a-z0-9-]{0,63}$/;

export function isCursor(value: string): boolean {
  return CURSOR.test(value);
}

export function isTourId(value: string): boolean {
  return TOUR_ID.test(value);
}

/**
 * Reads the response body of GET /api/v1/account/tours into a flat map,
 * discarding anything that is not a string value.
 *
 * Exported for its own test: the shape crosses a network boundary from a
 * column that clients write, and "it will be a flat object" is an
 * assumption, not a guarantee.
 */
export function parseTours(body: unknown): TourProgress {
  if (typeof body !== 'object' || body === null) {
    return NO_PROGRESS;
  }
  const tours = (body as { tours?: unknown }).tours;
  if (typeof tours !== 'object' || tours === null || Array.isArray(tours)) {
    return NO_PROGRESS;
  }
  const out: Record<string, string> = {};
  for (const [key, value] of Object.entries(tours as Record<string, unknown>)) {
    if (typeof value === 'string' && isTourId(key) && isCursor(value)) {
      out[key] = value;
    }
  }
  return out;
}

/**
 * This account's tour progress, or nothing recorded.
 *
 * Never throws and never rejects: no API configured, no token, a non-2xx,
 * a body in the wrong shape and a network failure all mean the same thing
 * to the caller — start the tour from the beginning — and distinguishing
 * them would only give the screen a decision it has no way to act on.
 */
export async function readTourProgress(
  baseUrl: string | undefined,
  token: string | null,
  fetchImpl: typeof fetch = fetch,
): Promise<TourProgress> {
  if (baseUrl === undefined || baseUrl === '' || token === null || token === '') {
    return NO_PROGRESS;
  }
  try {
    const response = await fetchImpl(`${baseUrl.replace(/\/$/, '')}/api/v1/account/tours`, {
      headers: { Authorization: `Bearer ${token}`, Accept: 'application/json' },
    });
    if (!response.ok) {
      return NO_PROGRESS;
    }
    return parseTours(await response.json());
  } catch {
    return NO_PROGRESS;
  }
}

/**
 * Records one cursor. Reports whether the API stored it, so the caller can
 * decide whether the browser's own copy is now the only one — it does not
 * throw, for the same reason the read does not.
 */
export async function writeTourProgress(
  baseUrl: string | undefined,
  token: string | null,
  tourId: string,
  cursor: string,
  fetchImpl: typeof fetch = fetch,
): Promise<boolean> {
  if (baseUrl === undefined || baseUrl === '' || token === null || token === '') {
    return false;
  }
  // Checked here as well as by the API. The values come from a page
  // script, and a request the server is certain to refuse is not worth
  // making — nor worth the log line it would produce on the other side.
  if (!isTourId(tourId) || !isCursor(cursor)) {
    return false;
  }
  try {
    const response = await fetchImpl(
      `${baseUrl.replace(/\/$/, '')}/api/v1/account/tours/${tourId}`,
      {
        method: 'PUT',
        headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
        body: JSON.stringify({ cursor }),
      },
    );
    return response.ok;
  } catch {
    return false;
  }
}

File: internal/account/tours.go
package account

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"log/slog"
	"net/http"
	"regexp"
	"slices"

	"github.com/google/uuid"

	platformhttp "github.com/Nomos-N4s/apivo-news/internal/platform/http"
)

// maxBodyBytes bounds every request body here. The largest legal one is a
// single short cursor, so this is orders of magnitude of headroom and
// exists only so a hostile body cannot be read into memory.
const maxBodyBytes = 4 << 10

// MaxToursPerAccount bounds how many tours one account may have progress
// for.
//
// The tour ids come from the front end, not from a table, which means an
// authenticated caller can PUT any id they invent. Without a ceiling the
// document grows until the row does, and the person paying for that is
// whoever next reads the account — the row is loaded on ordinary requests.
// A cap turns an unbounded write into a 409 that names the problem.
const MaxToursPerAccount = 32

// tourIDPattern is what may become a key in the tour_progress document.
// Narrow on purpose: these are identifiers from this product's own front
// end, and anything outside the shape is either a typo or somebody probing
// what the column will hold.
var tourIDPattern = regexp.MustCompile(`^[a-z][a-z0-9-]{0,63}$`)

// cursorPattern is what may become a value. The front end owns what a
// cursor MEANS — a step index into a list defined there — and this only
// refuses shapes the column should never hold. Four digits is a tour of
// ten thousand steps, which is not a tour.
var cursorPattern = regexp.MustCompile(`^(done|[0-9]{1,4})$`)

// TourStore reads and writes one account's tour progress.
type TourStore interface {
	// Tours returns the account's whole progress document. A missing
	// account reports ErrNoAccount.
	Tours(ctx context.Context, accountID uuid.UUID) (map[string]string, error)
	// SetTour records a cursor for one tour, returning false when the
	// account already holds MaxToursPerAccount other tours. A missing
	// account reports ErrNoAccount.
	SetTour(ctx context.Context, accountID uuid.UUID, tourID, cursor string) (bool, error)
}

// ErrNoAccount reports that the token resolved to an account id with no
// row behind it — deleted between authentication and this query. Distinct
// from a failed query, which leaves the question unanswered.
var ErrNoAccount = errors.New("account: no such account")

// Handler is this module's route table.
type Handler struct {
	log   *slog.Logger
	store TourStore
	auth  Authenticator
}

// NewHandler builds the account route table. The composition root mounts
// it under the prefix every pattern below shares.
func NewHandler(log *slog.Logger, store TourStore, auth Authenticator) http.Handler {
	h := &Handler{log: log, store: store, auth: auth}
	mux := http.NewServeMux()
	for pattern, handler := range h.routes() {
		mux.HandleFunc(pattern, handler)
	}
	// Every error under this prefix is problem+json, including paths
	// nobody wrote a handler for — the same convention the reader and
	// editorial modules hold, so the API has one error shape rather than
	// one per module plus ServeMux's text/plain in the corners.
	mux.HandleFunc("/api/v1/account/", h.handleUnrouted)
	return h.requireAccount(mux)
}

func (h *Handler) routes() map[string]http.HandlerFunc {
	return map[string]http.HandlerFunc{
		"GET /api/v1/account/tours":        h.readTours,
		"PUT /api/v1/account/tours/{tour}": h.writeTour,
	}
}

// Patterns lists this module's ServeMux patterns ("METHOD /path"), sorted,
// for the contract test in cmd. The catch-all is deliberately absent: it
// is the error convention for paths nobody serves, not an endpoint.
func Patterns() []string {
	h := &Handler{}
	patterns := make([]string, 0, len(h.routes()))
	for pattern := range h.routes() {
		patterns = append(patterns, pattern)
	}
	slices.Sort(patterns)
	return patterns
}

// handleUnrouted answers anything under the prefix that no route claims.
func (h *Handler) handleUnrouted(w http.ResponseWriter, r *http.Request) {
	platformhttp.Problem(w, http.StatusNotFound, "no such account endpoint: "+r.Method+" "+r.URL.Path)
}

// toursResponse is the body of GET /api/v1/account/tours.
type toursResponse struct {
	// Tours maps a tour id to its cursor. Always present, `{}` when the
	// account has started none — an absent field and an empty object would
	// mean the same thing and the client would have to handle both.
	Tours map[string]string `json:"tours"`
}

func (h *Handler) readTours(w http.ResponseWriter, r *http.Request) {
	tours, err := h.store.Tours(r.Context(), accountFrom(r.Context()).ID)
	switch {
	case errors.Is(err, ErrNoAccount):
		platformhttp.Problem(w, http.StatusNotFound, "this token authenticates an account that no longer exists")
		return
	case err != nil:
		h.internalError(w, r, "reading tour progress", err)
		return
	}
	if tours == nil {
		tours = map[string]string{}
	}
	h.writeJSON(w, r, http.StatusOK, toursResponse{Tours: tours})
}

// tourWrite is the body of PUT /api/v1/account/tours/{tour}.
type tourWrite struct {
	Cursor string `json:"cursor"`
}

func (h *Handler) writeTour(w http.ResponseWriter, r *http.Request) {
	tourID := r.PathValue("tour")
	if !tourIDPattern.MatchString(tourID) {
		platformhttp.Problem(w, http.StatusBadRequest, "a tour id is lower-case letters, digits and hyphens, starting with a letter")
		return
	}
	var body tourWrite
	if !decodeJSON(w, r, &body) {
		return
	}
	if !cursorPattern.MatchString(body.Cursor) {
		platformhttp.Problem(w, http.StatusBadRequest, `cursor must be "done" or a step number`)
		return
	}

	stored, err := h.store.SetTour(r.Context(), accountFrom(r.Context()).ID, tourID, body.Cursor)
	switch {
	case errors.Is(err, ErrNoAccount):
		platformhttp.Problem(w, http.StatusNotFound, "this token authenticates an account that no longer exists")
		return
	case err != nil:
		h.internalError(w, r, "recording tour progress", err)
		return
	}
	if !stored {
		platformhttp.Problem(w, http.StatusConflict, "this account already tracks the maximum number of tours")
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// decodeJSON decodes the request body into dst, answering the 400 itself
// when the body is not the expected shape. Unknown fields are rejected: a
// misspelled field silently ignored would read as acceptance.
func decodeJSON(w http.ResponseWriter, r *http.Request, dst any) bool {
	dec := json.NewDecoder(http.MaxBytesReader(w, r.Body, maxBodyBytes))
	dec.DisallowUnknownFields()
	if err := dec.Decode(dst); err != nil {
		platformhttp.Problem(w, http.StatusBadRequest, "request body is not valid JSON for this endpoint: "+err.Error())
		return false
	}
	// Requiring io.EOF rather than !dec.More() rejects both a second
	// document and trailing syntax errors; More() answers the narrower
	// question "is another VALUE coming?", so a stray `}` reads as no.
	if err := dec.Decode(&json.RawMessage{}); !errors.Is(err, io.EOF) {
		platformhttp.Problem(w, http.StatusBadRequest, "request body must contain a single JSON document")
		return false
	}
	return true
}

func (h *Handler) writeJSON(w http.ResponseWriter, r *http.Request, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(body); err != nil {
		h.log.WarnContext(r.Context(), "writing account response", "error", err)
	}
}

// internalError logs the failure and answers an opaque 500: internals are
// for the log, never the wire.
func (h *Handler) internalError(w http.ResponseWriter, r *http.Request, doing string, err error) {
	h.log.ErrorContext(r.Context(), doing, "error", err)
	platformhttp.Problem(w, http.StatusInternalServerError, "")
}
